devicemapper: Implement deferred deletion capability

rhvgoyal opened the pull request:
Fixes #16380 

This patch series implements devicemapper deferred device deletion capability. By default it is turned off and one can enable it by passing parameter "--storage-opt dm.use_deferred_deletion=true". This requires deferred removal to be turned on. In fact without that deferred deletion does not make much sense. So pass following to docker daemon.

"--storage-opt dm.use_deferred_deletion=true --storage-opt dm.use_deferred_removal=true"

rhvgoyal commented:
@vbatts @rhatdan Please have a look.

rhatdan commented:
enableDeferredRemoval        = false
+	enableDeferredDeletion       = false

Do we need both flags or should DeferredRemoval imply Deferred Deletion?  Is there ever a case where you would not want both?

rhvgoyal commented:
We have have deferred removal without deferred deletion. So we need both the flags. But deferred deletion without deferred removal might not make much sense. So one suggestion mike snitzer gave that when user specifies use_deferred_deletion=true, then automatically set user_deferred_removal=true internally. 

As I can't see a use case of deferred deletion without deferred removal enabled, I will make this change.

rhatdan commented:
My point being isn't using deferred removal without deferred deletion just opening you up to a bug where the deletion fails?  Why would you not want to use deferred deletion all the time?

rhvgoyal commented:
There are couple of reasons.

- There might be bugs in this new logic. So people might want to disable deferred device deletion for the time being.

- Using deferred deletion leads to deletion of container root directory and that in turn leads to deletion of that container rootfs mount point from all the mount namespaces. It is possible that some people might like old behavior where container deletion fails instead.

Not being able to delete a container because it is busy in some namespace is not a bug necessarily. We are offering two behaviors for device deletion now. Synchronous and asynchronous and depending on what behavior one wants, users can choose either.

LK4D4 commented:
18 commits is too much for review. It should be 2-3 at most.

rhvgoyal commented:
@LK4D4 

Ok, I will shrink it down to 2-3 commits. I like it to keep them nicely broken for my own sanity so that I can look at small pieces of code and figure out why a patch is required. Keep cleanups and bug fixes and code re-organization in separate patches. That way one can look at the core piece.

But, as docker project rather prefers a single large patch as compared to broken down smaller pieces,  will do that.

rhvgoyal commented:
@LK4D4 

If you are looking for a single patch view, You just have to click on "Files Changed" tab and then you can review it. Right?

So breaking down into smaller logical pieces seems to give you best of both the worlds. Those who prefer looking at smaller patches, they can review it per commit and those who want to look at single
large patch they can just go to "Files Changed" view?

rhvgoyal commented:
I have written very nice commit logs with each patch. And these will be useful in future too when somebody is wondering why a particular change has been done. Squishing everything into 2-3 commits will mean that you will lose all these changelogs.

snitm commented:
@rhvgoyal you are the one who needs to be able to maintain this code.  Don't compromise on what enables you to work most effectively, now and in the future should you need to revisit the "why?"s of your change, purely because some reviewer(s) have some bizarre notion of patchset granularity.  Keep up the good work! ;)

tiborvass commented:
Ping @vbatts @rhatdan for code review/approval

rhvgoyal commented:
Pushed patches again. Fixed couple of error messages and also added documentation to docker man page.

rhatdan commented:
+		return fmt.Errorf("devmapper: Can't activate device %v as it is makred for deletion", info.Hash) 

Typo marked

rhatdan commented:
+// Cleanup the deleted devices during startup. It assumes that all the
+// all devices have been loaded into hash table.
+// Should be called with devices.Lock() held. Will drop the lock for
+// deivice deletion and return with lock acquired.

Spelling and typos

rhatdan commented:
```
	if enableDeferredDeletion && !devices.deferredRemove {
+		logrus.Infof("devmapper: Deferred deletion can not be enabled as deferred removal is not enabled.")
+	}
+
+	if devices.deferredRemove && enableDeferredDeletion {
+		logrus.Debugf("devmapper: Deferred deletion support enabled.")
+		devices.deferredDelete = true
+	}
```

Should these be combined into a single fblock.  Still not convinced that this should not be an error situation.  DeferredRemove false DeferredDeletion true.

rhvgoyal commented:
@rhatdan 

So for the case of deferred_remove=false deferred_deletion=true, right now I am just giving warning and keeping deferred_deletion=false. 

So you are suggesting that instead error out saying that deferred deletion can't be used? I think that's doable if that provides better semantics.

rhatdan commented:
I am thinking most likely this is a configuration problem by the admin, so why continue.

rhvgoyal commented:
@rhatdan 

I fixed typos. Also changed the code so that it will error out if deferred deletion is requested but it can not be enabled.

We had the similar behavior for deferred removal code. If user requested it but either kernel does not support it or libdm does not support it we will not enable it and docker will continue to run. But now I have changed it and if user requested deferred removal and it can not be enabled, then docker will fail.

rhvgoyal commented:
Something went wrong. It is not seeing my commits. Trying to fix it.

rhvgoyal commented:
@rhatdan 

I have changed the patches to move the discard check location. PTAL.

rhatdan commented:
LGTM  

@vbatts PTAL

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1503,7 +1718,92 @@ func (devices *DeviceSet) DeleteDevice(hash string) error {
+	// or has not been Put() properly. Fail device deletion.
+
+	if info.mountCount != 0 {
+		return fmt.Errorf("devmapper: Can't delete device %v as it is still mounted. mntCount=%v", info.Hash, info.mountCount)
```
is this going to be different functionality than currently?
Or is this just returning earlier from what would be an EBUSY?

rhvgoyal replied:
This is more of a saftey check to make sure that docker has done proper number of Get()/Put() so that mount count is zero. If not, then it is most likely  a upper layer bug where a Get() was done but corresponding Put() has not been done and higher layer is trying to delete the device. So instead of scheduling a deferred delete now, it will force higher layers to first do all corresponding Put() and then call into delete device logic. If device is busy even now as it leaked into other namespaces, then deferred deletion will be scheduled.

vbatts replied:
understood

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -542,7 +563,48 @@ func (devices *DeviceSet) migrateOldMetaData() error {
 	return nil
 }
 
+// Cleanup deleted devices during startup. It assumes that all the
```
technically not _just_ startup

vbatts replied:
sorry. I thought I was commenting on the thread function

rhvgoyal replied:
Right. This comment can be changed now. Initially I was doing cleanup only during startup and that's what this comment reflects. Later I also introduced a thread to do the cleanup. I will update this comment.

vbatts replied:
cool. thanks.

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1503,7 +1718,92 @@ func (devices *DeviceSet) DeleteDevice(hash string) error {
+		logrus.Debugf("devmapper: Calling cleanupDeletedDevices().\n")
+		err := devices.cleanupDeletedDevices()
+		if err != nil {
+			logrus.Errorf("devmapper: cleanupDeletedDevices() failed.")
```
given that this may fail, and the `devices.deletionWorkerRunning` would not be reset to `false`, then this thread would never be respawned until daemon restart, correct?

rhvgoyal replied:
When error happens, I use "break" to break out of for loop and then set devices.deletionWorkerRunning=false. So next time thread should still run. Do you see a return somewhere upon error?

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1503,7 +1718,92 @@ func (devices *DeviceSet) DeleteDevice(hash string) error {
+			break
+		}
+	}
+	devices.deletionWorkerRunning = false
```
oh wait, I see. It would break to here, and return.
Nevermind me.

rhvgoyal replied:
Ah, you noticed it. cool.

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1503,7 +1718,92 @@ func (devices *DeviceSet) DeleteDevice(hash string) error {
+	go devices.cleanupDeletedDevicesThread()
+}
+
+func (devices *DeviceSet) cancelDefDeletion() {
```
might you comment on this function the assurance that this needs to provide. With the sleeps and all, you obviously are guarding something.

rhvgoyal replied:
Will do. Right now it is a blocking function which waits for worker thread to stop.

vbatts replied:
thanks.

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1503,7 +1718,92 @@ func (devices *DeviceSet) DeleteDevice(hash string) error {
+		devices.Unlock()
+		// Sleep for 10 miliseconds and check again
+		time.Sleep(time.Millisecond * 10)
+	}
```
is there a situation that this loop would run indefinitely? This would be a blocking operation. Perhaps just a sane number of iterations before bailing?

rhvgoyal replied:
If there is a bug, this can run indefinitely. Not sure what's the sane number where we can say stop. Thought it is not a big concern at this point as it is called during docker stop only and if Shutdown() does not finish in certain time (10 second), then all threads are killed anyway.

vbatts replied:
ah yeah. This seems fine for now then

vbatts commented:
all-in-all, LGTM. A couple of questions marked though.

@rhvgoyal I love the clarity in your commits.

rhvgoyal commented:
@vbatts I have taken care of your comments. Updated the description of two functions. PTAL.

vbatts commented:
LGTM. and janky _had_ passed, but appears to be having issues.

vbatts commented:
ping @thaJeztah @moxiegirl

moxiegirl commented on man/docker.1.md:
```diff
@@ -425,6 +425,26 @@ device.
 
 Example use: `docker daemon --storage-opt dm.use_deferred_removal=true`
 
+#### dm.use_deferred_deletion
```
@rhvgoyal looks useful. A few things, this section seems to repeat itself.  It would be more useful to restructure it a bit. Also, I think it is useful to include the error users encounter if this is *not* enabled.

Also, you need to add this same section to the `docs/reference/commandline/daemon.md` file

-->  https://gist.github.com/moxiegirl/b8bbf5b4c74ad2f35e9c

#### dm.use_deferred_deletion

Enables use of deferred device deletion for thin pool devices. By default, thin pool device deletion is synchronous. Before a container is deleted, the Docker daemon removes any associated devices. If the storage driver cannot remove a device, the container deletion fails and the daemon returns:

```
Error deleting container: Error response from daemon: Cannot destroy container
```

To avoid this failure, enable both deferred device deletion and deferred device
removal on the daemon:

`docker daemon --storage-opt dm.use_deferred_deletion=true --storage-opt
dm.use_deferred_removal=true`

With these two options enabled, if a device is busy when the driver is deleting
a container, the driver marks the device as deleted. Later, when the device
isn't in use, the driver deletes it.

You should enable deferred device deletion in situations where you expect to
mount the directory of one container into another.  In these situations, a mount
point can leak into multiple namespaces causing the device to be busy.

rhvgoyal replied:
@moxiegirl Above documentation looks good. I will modify my patches and also add the same documentation in docs/reference/commandline/daemon.md file. I realize that documentation for
deferred removal is not present in daemon.md file, so I will add that too.

I don't understand the last point and that is recommending to use deferred deletion when mounting the directory of one container into another container. How does one do that? Is it about creating containers which explicitly export volumes and using those volumes in other containers.

I think I want to recommend the deferred deletion in general case. This feature primarily helps with unintentional device leak across mount namespaces. Where docker knows that it is sharing mount points explicitly, we should not allow container removal and block it at higher layers.

This feature is primarily useful when docker thinks that device/mount point is not being used by docker or any of the containers and it is safe to remove container. But there is a chance that device/mount point has leaked into other mount namespaces because somebody did "unshare -m" or because people did special volume mounts (-v /:/host). So this feature is is useful in those scenarios.

moxiegirl replied:
@rhvgoyal If it should be on in the general case why isn't it enabled by default?

The last point was my attempt to explain to the reader when to use this feature. You can also try:

"You should enable deferred device deletion on the device mapper if you regularly mount the directory of one container into another container.  This feature also helps when unintentional device leaks across mount namespaces are possible. For example, because a user did `unshare -m` or a special volume mount (`-v `/`:/host`). "

rhvgoyal replied:
@moxiegirl 

You have a good point that why this feature is not enabled by default. And only reason I did not do it is that it is very new feature. I will enable it by default in fedora. Once I have confidence that it is working reasonably well (based on fedora usage), I want to propose enabling it by default docker upstream.

So by the way, how do I "mount directory of one container into another container"?

moxiegirl commented on man/docker-daemon.8.md:
```diff
@@ -297,6 +297,28 @@ device.
+Docker daemon removes any associated devices. If the storage driver can not
+remove a device, the container deletion fails and daemon returns.
+
+Error deleting container: Error response from daemon: Cannot destroy container
```
@rhvgoyal You need to add tics around these two lines to put them in code font.

rhvgoyal replied:
Will do.

moxiegirl commented on man/docker-daemon.8.md:
```diff
@@ -297,6 +297,28 @@ device.
+To avoid this failure, enable both deferred device deletion and deferred
+device removal on the daemon.
+
+docker daemon --storage-opt dm.use_deferred_deletion=true --storage-opt dm.use_deferred_removal=true
```
@rhvgoyal You need to add tics around these two lines to put them in monospace:

rhvgoyal replied:
will do.

moxiegirl commented on docs/reference/commandline/daemon.md:
```diff
@@ -367,6 +367,46 @@ options for `zfs` start with `zfs`.
+    the Docker daemon removes any associated devices. If the storage driver
+    can not remove a device, the container deletion fails and daemon returns.
+
+    Error deleting container: Error response from daemon: Cannot destroy container
```
@rhvgoyal You need to add tics around these two lines to put them in monospace:

rhvgoyal replied:
@moxiegirl Will do.

moxiegirl commented on docs/reference/commandline/daemon.md:
```diff
@@ -367,6 +367,46 @@ options for `zfs` start with `zfs`.
+    To avoid this failure, enable both deferred device deletion and deferred
+    device removal on the daemon.
+
+    docker daemon --storage-opt dm.use_deferred_deletion=true --storage-opt dm.use_deferred_removal=true
```
@rhvgoyal You need to add tics around these two lines to put them in monospace:

rhvgoyal replied:
@moxiegirl Will do.

calavera commented:
Hi all,

I just moved this pull request to `code-review`. Only one docker maintainer LGMTed this pull request. This change is significantly big to require at least two maintainers to give it the :+1:.

vbatts commented:
@calavera ah you're right. I thought I saw another one, but it was @rhatdan

rhatdan commented:
Yes I would not trust him.  :^)

calavera commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1460,41 +1600,111 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
+	logrus.Debugf("devmapper: Marking device %s for deferred deletion.", info.Hash)
+
+	info.Deleted = true
+	devices.nrDeletedDevices++
```
If the call to `devices.saveMetadata` fails two lines below, this value will become incorrect. This value should either be incremented after calling `devices.saveMetadata` or be decremented when there is an error calling `devices.saveMetadata`.

rhvgoyal replied:
Ok, will change it.

calavera commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1460,41 +1600,111 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
 
-	if err := devices.openTransaction(info.Hash, info.DeviceID); err != nil {
-		logrus.Debugf("Error opening transaction hash = %s deviceID = %d", "", info.DeviceID)
+	if err := devices.closeTransaction(); err != nil {
```
This `closeTransaction` is never called if any of the errors before interrupts the function. Is this correct?

rhvgoyal replied:
Dan Walsh had also pointed out this. I think it is not very bad in the sense that if we don't close transaction, that means if docker crashes now, it will try to delete this device when docker restarts and that's what we wanted to begin with anyway.

It probably will be a problem when we try to delete the container later and it will return error as device has already been deleted during recovery. So yes, it is a good idea to close the transaction in case of error. 

Will fix it.

calavera commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -395,27 +425,18 @@ func (devices *DeviceSet) deviceFileWalkFunction(path string, finfo os.FileInfo)
-		return fmt.Errorf("Error loading device metadata file %s", hash)
+	// Include deleted devices also as cleanup delete device logic
+	// will go through it and see if there are any deleted devices.
+	if _, err := devices.lookupDevice(hash); err != nil {
```
you mentioned in the comment above the signature of `lookupDevice` that this function should always be called after locking the deviceSet. There is nothing in `deviceFileWalkFunction` or the call stack that locks the deviceSet, at least as far as I can tell. Should we call here to `lookupDeviceLock`? why do we need a separated function if that function should always be called with a lock held?

rhvgoyal replied:
initMetaData() takes the devices.Lock().

initMetaData --> loadDeviceFilesOnStart() ----> deviceFileWalkFunction ---> lookupDevice()

There are couple of cases where I have lock already held and don't need a locked version
of lookupDevice(), hence I need both the versions.

calavera replied:
sounds good.

calavera commented:
@rhatdan, @vbatts no worries. It was in my queue to review and I was surprised that it moved to docs review already. I have a few comments, one might be about a bug :smile_cat:

calavera commented:
@rhvgoyal I really like all the information you added to your commits, it's very helpful. However, I agree with @LK4D4 that all the commits will need to be squashed before merging this to master. This doesn't mean that I want you to remove all those comments, don't get me wrong. I'd like to keep all that information in the resulting commit.

I agree that @rhvgoyal should maintain this code, since he wrote it, and he should be able to understand what he did. But there are a group of people that should be able to work efficiently with this code. That might mean to move the code around, or it could imply to be able to revert it in the case we found regressions/bugs. The best way to do that is to have as less commits as possible.

@snitm This is not a random request by a random reviewer. This is the commit policy we follow along the whole project. I agree that some of the changes could stand out by themselves, but unfortunately all these commits are too much coupled to each other, and it's a single feature after all. 

Finally, I hope you can stop encouraging people from ignoring the project maintainers comments. At the end of the day, the group of people working on the project in a daily bases are the ones that have to deal with this code. Not only this specific code but also other requests from other users and our own roadmap. Having consistency in the way people contribute to the project helps us be more efficient in our job. Encouraging people to do whatever they want does not help us.

snitm commented:
@calavera I really don't want to derail progress on the review of @rhvgoyal's code with a tangent about a docker contributor's changes should or shouldn't be managed.  It just strikes me as _very_ odd that there is some perceived benefit to folding a series of commits, with separate headers, into a single commit.  In the end my point is: there is value in seeing how the code evolved and to preserve that evolution.  If a revert is needed (or the changes across all related commits need to be inspected) in bulk there is always:
git diff commit1^..commentN
git diff commit1^..commentN | patch -p1 -R && git commit -a
Having the project maintainers impose a rigid "one commit per feature" is, IMHO, needless redtape but whatever floats your docker boat.. but ignore me, as you guys were with this review! ;)

rhvgoyal commented:
@calavera 

Once I have got LGTM from code review perspective, and from doc perspective, I will reduce these commits to fewer commits. and merge their commit logs too.

calavera commented on pkg/devicemapper/devmapper.go:
```diff
@@ -750,7 +750,11 @@ func DeleteDevice(poolName string, deviceID int) error {
 
+	dmSawBusy = false
 	if err := task.run(); err != nil {
+		if dmSawBusy == true {
```
this could also be `if dmSawBusy {`

rhvgoyal replied:
Right. That's shorter form. Will change it.

rhvgoyal commented:
@calavera @moxiegirl I have taken care of your comments and pushed patches again. PTAL.

moxiegirl commented:
Thanks @rhvgoyal LGTM

calavera commented:
I've noticed that some errors start with the prefix `devmapper:` but others not. I'd like to be consistent, so either we add the prefix to all of them or we remove it from the ones we already have. Unfortunately, I think that problem is not specific to this pull request. It would be great if we could consolidate those messages.

Otherwise LGTM.

calavera commented:
Moved back to docs review :tada:

rhvgoyal commented:
@calavera I noticed that too. I have opened a separate issue to track it. I agree it is a good idea to make all messages have same prefix. Once this PR gets merged, I will open another one to fix it.

https://github.com/docker/docker/issues/16667

rhvgoyal commented:
@calavera @vbatts Ok, I have merged the patches and came up 3 patches for merging.

I feel docker team should rethink about their policy of limiting patches to 2-3 max. It works well for small fixes but if somebody is trying to do a proper feature, then it just forces one to write a big blob of code in a single patch and I really I can't see how that is a good thing overall.

We have been following the policy of broken down patches in linux kernel and it works very well. So there is a precedent for sure. There needs to be a *very* good reason to defend this docker policy. So far I have not seen a good argument to support the policy.

rhvgoyal commented:
Ok, I don't want to squash all the changes into smaller patches. So based on the discussion on #docker-dev, I am going to route smaller set of changes into individual PRs.

First of the series of PR is here. Please have a look.

https://github.com/docker/docker/pull/16683

LK4D4 commented:
@rhvgoyal need rebase
I'll review in the meantime

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -555,7 +576,47 @@ func (devices *DeviceSet) migrateOldMetaData() error {
+	var deletedDevices []*devInfo
+
+	for _, info := range devices.Devices {
+		if info.Deleted == false {
```
if `!info.Deleted`, for other places too.

rhvgoyal replied:
Will do.

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -141,7 +146,9 @@ type Status struct {
 	UdevSyncSupported bool
 	// DeferredRemoveEnabled is true then the device is not unmounted.
-	DeferredRemoveEnabled bool
+	DeferredRemoveEnabled      bool
```
What the difference betwee `RemoveEnalbed` and `DeleteEnabled`?

rhvgoyal replied:
Removal is device deactivation. So it basically removes the device like /dev/dm-1. Deletion is actually removal of thin device from the thin pool. Removal does not delete the device from thin pool and once can activate it again later. But once the deletion is done, device is gone for ever and can't be activated again.

rhvgoyal replied:
So RemoveEnabled tells whether deferred device deactivation is enabled or not. And DeleteEnabled tells whether deferred deletion is enabled or not. "Remove" and "Delete" are pretty close terms and it is very confusing. In the context of device mapper, they refer to different operations.

LK4D4 replied:
Deserves comment here I think, I dunno why golint isn't complaining, btw.

rhvgoyal replied:
Ok, I will put a comment to highlight difference between Deferred Remove and Deferred Delete operation.

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -388,6 +413,11 @@ func (devices *DeviceSet) deviceFileWalkFunction(path string, finfo os.FileInfo)
 	}
 
+	if finfo.Name() == transactionMetaFile {
+		logrus.Debugf("Skipping file %s", path)
```
maybe "skipping transaction metafile"

rhvgoyal replied:
This already got merged as part of other PR. I will take care of it when I am doing a general cleanup to bring all the messages to a consistent prefix as part of separate PR.

LK4D4 commented:
@rhvgoyal I thinking about cleanup deferred loop. Goroutines is extremely cheap in go, so maybe we can start one goroutine which will wake up every 30 seconds(with `time.Tick`) and check if there are some stuff for deleting. This will simplify code and logic overall, I think it might be even good idea to split generic package for that task later.
WDYT?

rhvgoyal commented:
@LK4D4 So we want to a continuous running thread and get rid of logic of scheduling the thread. I think that is doable. And that should get rid of need of calling scheduleDefDeletion() once there are deleted devices. We will have to start the worker thread during startup.

But we will still need the capability to stop the thread during Shutdown(). That means we still will have to design worker thread in such a way so that it sets up a channel with parent and listens for any signal to stop the thread. When Shutdown() is called, we want to make sure worker thread has exited.

I have taken care of rest of the comments. I will push those patches for now and then work on this alternate proposal and see how does it look.

LK4D4 commented:
@rhvgoyal Yeah, you can `Stop()` `time.Ticker`, this will stop goroutine. You can store ticker in `DeviceSet`, should be nice and clean. At least looks now, `for ticker.C` in GC goroutine and `ticker.Stop()` in `Shutdown()`

rhvgoyal commented:
@LK4D4 I have rebased the patches. Out of 6 got merged in other PRs and I have merged one patch in existing patches. So we are left with 11 now. I think I will open 1-2 more PRs and get to reduce the number further.

I have implemented your suggestion of using time.Ticker and this seems to be working well. Please have a look at the last patch.

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -553,6 +580,62 @@ func (devices *DeviceSet) migrateOldMetaData() error {
+	}
+
+	logrus.Debugf("devmapper: Worker to cleanup deleted devices started")
+	for _ = range devices.deletionWorkerTicker.C {
```
This is unnecessary since go1.3 IIRC, you can just write `for range devices.deletionWorkerTicker.C`

rhvgoyal replied:
Yep. That worked. Changed.

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -553,6 +580,62 @@ func (devices *DeviceSet) migrateOldMetaData() error {
+		deletedDevices = append(deletedDevices, info)
+	}
+
+	// Delete the delete devices. DeleteDevice() first takes the info lock
```
Looks like should be `Delete the deleted...`

rhvgoyal replied:
Fixed.

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -553,6 +580,62 @@ func (devices *DeviceSet) migrateOldMetaData() error {
+// loaded in the hash table. Should be called with devices.Lock() held.
+// Will drop the lock for device deletion and return with lock acquired.
+func (devices *DeviceSet) cleanupDeletedDevices() error {
+	var deletedDevices []*devInfo
```
Unlike in C in go it's better to decalre vars closer to their usage. So it should be after zero-check.

rhvgoyal replied:
Fixed.

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -553,6 +580,62 @@ func (devices *DeviceSet) migrateOldMetaData() error {
+		return nil
+	}
+
+	for _, info := range devices.Devices {
```
btw you have same loop here and in countDeletedDevices, but you use different rule for condition(continue on false vs action on true), I'm not against any, but it should be same I think.

rhvgoyal replied:
Done. Made both the loops same.

LK4D4 commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -553,6 +580,62 @@ func (devices *DeviceSet) migrateOldMetaData() error {
+
+	logrus.Debugf("devmapper: Worker to cleanup deleted devices started")
+	for _ = range devices.deletionWorkerTicker.C {
+		devices.Lock()
```
Just move locking inside function, it looks weird and unsafe to lock in one function and unlock in another.

rhvgoyal replied:
Done. Initially I was calling cleanupDeletedDevice() from initMetaData() where devices lock was already held. So I could not take lock inside function. Now this function has been moved in a separate thread, so I can take lock inside function. Changed the last patch.

LK4D4 commented:
@rhvgoyal gofmt complaining.
Ticker stuff looks cool for me, but there are still some unrelated commits in PR.
Also, I found that locking logic is extremely tangled in devicemapper package. Probably it's trying to be more granular than it should be and that brings complexity of code and increase chance of introducing deadlock. But that's definitely not a problem of this PR.

rhvgoyal commented:
@lk4d4 I will look into gofmt issue. Surprisingly on my machine gofmt is not flagging anything wrong. May be it has something to do with gofmt version.

I have generated another PR to push 3 commits. I have I will have to generate 1-2 more PRs to make sure that only deferred deletion related patches remain in this PR.

I will look into simplifying locking even further. But as you say, that's subject for another PR after these changes get merged.

LK4D4 commented:
@rhvgoyal I think we run gofmt with -s argument. Version can be problem too.

rhvgoyal commented:
@LK4D4 I have taken care of your review comments and pushed again. I hope go formatting issues have been taken care of. I have run "gofmt -s -w". If not I will have to look deeper.

I have merged first three patches and they are lined up to be merged in a separate PR.

https://github.com/docker/docker/pull/16727

PTAL.

rhvgoyal commented:
PR #16727 got merged. I dropped top level patch and rebased patches and pushed again. Now I am left with 8 patches in the series.

I have created another PR #16799 for top 3 patches to break down setupBaseImage() function into smaller pieces. Once that gets merged I will drop top 3 patches. Hopefully then rest of the 5 patches can be merged as part of this PR.

LK4D4 commented:
@rhvgoyal btw can you reword last commit a little? It's actually not a thread, but goroutine.

rhvgoyal commented:
@lk4d4 will do.

rhvgoyal commented:
In last patch, renamed references to `thread` with `goroutine`

vbatts commented:
rebase please

rhvgoyal committed:
devmapper: Provide option to enabled deferred device deletion

Provide a command line option dm.use_deferred_deletion to enable deferred
device deletion feature. By default feature will be turned off.

Not sure if there is much value in deferred deletion being turned on
without deferred removal being turned on. So for now, this feature can
be enabled only if deferred removal is on.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal committed:
devmapper: construct used device ID map from device Hash map

Currently during startup we walk through all the device files and read
their device ID and mark in a bitmap that device id is used.

We are anyway going through all device files. So we can as well load all
that data into device hash map. This will save us little time when
container is actually launched later.

Also this will help with later patches where cleanup deferred device
wants to go through all the devices and see which have been marked for
deletion and delete these.

So re-organize the code a bit.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal committed:
devmapper: Implement a goroutine to cleanup deleted devices

Start a goroutine which runs every 30 seconds and if there are deferred
deleted devices, it tries to clean those up.

Also it moves the call to cleanupDeletedDevices() into goroutine and
moves the locking completely inside the function. Now function does not
assume that device lock is held at the time of entry.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal committed:
devmapper: Keep track of number of deleted devices

Keep track of number of deleted devices and export this information through
"docker info".

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal committed:
devmapper: Implement deferred deletion functionality

Finally here is the patch to implement deferred deletion functionality.
Deferred deleted devices are marked as "Deleted" in device meta file. 

First we try to delete the device and only if deletion fails and user has
enabled deferred deletion, device is marked for deferred deletion.

When docker starts up again, we go through list of deleted devices and
try to delete these again.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@lk4d4 @vbatts Now only last 5 patches are remaining in this PR. They all are related to this deferred device deletion feature. I am hoping that this PR can now be merged as it is without having the need to break it down further. PTAL.

thaJeztah commented:
thanks so much @rhvgoyal, docs LGTM (and Mary already LGTM'd above https://github.com/docker/docker/pull/16381#issuecomment-144191542)

@LK4D4 @vbatts if the patch-set looks good to you, I think it's ready to merge

LK4D4 commented:
LGTM

thaJeztah commented:
:tada: thanks!
